File: native/ios/Auspice/Screens/Almanac.swift
import SwiftUI

// MARK: - What the rules hand back

struct AlmanacDay: Codable {
    struct Lunar: Codable {
        let text: String
        let yearGanZhi: String
        let monthGanZhi: String
        let dayGanZhi: String
        let zodiac: String
    }

    struct SolarTermRef: Codable {
        let name: String
        let date: String
    }

    struct Mansion: Codable {
        let name: String
        let animal: String
        let direction: String
        let beast: String
    }

    struct Spirit: Codable {
        let name: String
        let road: String
        let luck: String
    }

    struct Hour: Codable, Identifiable {
        let ganzhi: String
        let range: String
        let spirit: String
        let lucky: Bool
        var id: String { ganzhi + range }
    }

    struct Judgement: Codable {
        struct Basis: Codable {
            let zh: String
            let en: String
        }
        let activity: String
        let verdict: String
        let matched: String?
        let basis: Basis
    }

    let date: String
    let lunar: Lunar
    let solarTerm: String?
    let nextSolarTerm: SolarTermRef
    let yi: [String]
    let ji: [String]
    let clash: String
    let harmDirection: String
    let dayOfficer: String
    let mansion: Mansion
    let spirit: Spirit
    let auspicious: [String]
    let inauspicious: [String]
    let pengzu: [String]
    let hours: [Hour]
    let standing: String
    let luckyHours: [Hour]
    let judgement: Judgement?
}

struct AlmanacActivity: Codable, Identifiable {
    struct Name: Codable {
        let zh: String
        let en: String
    }
    let id: String
    let name: Name
    let terms: [String]
}

/// Loads a day, and the list of undertakings, from the shared rules.
@Observable
final class AlmanacStore {
    static let shared = AlmanacStore()
    var day: AlmanacDay?
    var activities: [AlmanacActivity] = []
    var activity: String? = PracticeStorage.read("almanac.activity", default: Optional<String>.none) { didSet { PracticeStorage.write(activity, key: "almanac.activity") } }
    var date: Date = PracticeStorage.read("almanac.date", default: Date()) { didSet { PracticeStorage.write(date, key: "almanac.date") } }
    var error: String?

    private static let iso: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        formatter.calendar = Calendar(identifier: .gregorian)
        return formatter
    }()

    func load() {
        do {
            if activities.isEmpty {
                activities = try Engines.shared.evaluate("almanac.activities", as: [AlmanacActivity].self)
            }
            var input: [String: Any] = ["date": Self.iso.string(from: date)]
            if let activity { input["activity"] = activity }
            day = try Engines.shared.evaluate("almanac.day", input, as: AlmanacDay.self)
            error = nil
        } catch {
            self.error = l("This reading could not be computed. Please try again.")
        }
    }

    func shift(_ days: Int) {
        date = Calendar.current.date(byAdding: .day, value: days, to: date) ?? date
        load()
    }
}

// MARK: - The screen

struct AlmanacScreen: View {
    @State private var store = AlmanacStore.shared

    var body: some View {
        ScreenScaffold(
            eyebrow: t("practice.almanac"),
            title: t("almanac.title"),
            tagline: t("almanac.tagline")
        ) {
            dayPicker
            if let error = store.error {
                Panel(title: t("common.notComputed")) {
                    Text(error).font(Typeface.serif(16)).foregroundStyle(Palette.inkSoft)
                }
            }
            if let day = store.day {
                standingPanel(day)
                activityPanel(day)
                listsPanel(day)
                hoursPanel(day)
                tablesPanel(day)
                ExplainReading(result: day).id(day.date)
            }
        }
        .onAppear { if store.day == nil { store.load() } }
    }

    private var dayPicker: some View {
        HStack {
            Button { store.shift(-1) } label: {
                Image(systemName: "chevron.left").frame(width: 44, height: 44)
            }
            .foregroundStyle(Palette.inkSoft)
            Spacer()
            VStack(spacing: 2) {
                Text(store.date, format: .dateTime.weekday(.wide).day().month(.wide))
                    .font(Typeface.display(18))
                    .foregroundStyle(Palette.ink)
                if let lunar = store.day?.lunar.text {
                    Text(lunarDateText(lunar)).font(Typeface.serif(15)).foregroundStyle(Palette.inkMute)
                }
            }
            Spacer()
            Button { store.shift(1) } label: {
                Image(systemName: "chevron.right").frame(width: 44, height: 44)
            }
            .foregroundStyle(Palette.inkSoft)
        }
        .padding(.vertical, 2)
    }

    private func standingPanel(_ day: AlmanacDay) -> some View {
        Panel {
            HStack(alignment: .firstTextBaseline, spacing: 10) {
                Text(standingWord(day.standing))
                    .font(Typeface.display(26))
                    .foregroundStyle(standingColour(day.standing))
                Text(l(day.lunar.dayGanZhi))
                    .font(Typeface.serif(18))
                    .foregroundStyle(Palette.inkSoft)
                Spacer()
                Text(l(day.dayOfficer))
                    .font(Typeface.sans(13, weight: .bold))
                    .foregroundStyle(Palette.gold)
                    .padding(.horizontal, 10)
                    .padding(.vertical, 6)
                    .background(Capsule().fill(Palette.goldSoft))
            }
            Text(standingLine(day))
                .font(Typeface.serif(17))
                .foregroundStyle(Palette.inkSoft)
                .fixedSize(horizontal: false, vertical: true)
        }
    }

    private func activityPanel(_ day: AlmanacDay) -> some View {
        Panel(title: t("almanac.canIDoIt")) {
            FlowRow(spacing: 8) {
                ForEach(store.activities) { item in
                    Button {
                        store.activity = store.activity == item.id ? nil : item.id
                        store.load()
                    } label: {
                        HStack(spacing: 6) {
                            Text(l(item.name.en))
                        }
                        .font(Typeface.sans(14, weight: .semibold))
                        .padding(.horizontal, 14)
                        .frame(minHeight: 40)
                        .foregroundStyle(store.activity == item.id ? Palette.ink : Palette.inkSoft)
                        .background(
                            Capsule().fill(store.activity == item.id ? Palette.goldSoft : Color.white.opacity(0.04))
                        )
                        .overlay(
                            Capsule().strokeBorder(
                                store.activity == item.id ? Palette.goldLine : Palette.line,
                                lineWidth: 1
                            )
                        )
                    }
                    .buttonStyle(.plain)
                }
            }
            if let judgement = day.judgement {
                VStack(alignment: .leading, spacing: 6) {
                    Text(verdictWord(judgement.verdict))
                        .font(Typeface.display(22))
                        .foregroundStyle(verdictColour(judgement.verdict))
                    Text(judgementLine(judgement, day))
                        .font(Typeface.serif(17))
                        .foregroundStyle(Palette.inkSoft)
                        .fixedSize(horizontal: false, vertical: true)
                }
                .padding(.top, 4)
                .transition(.opacity.combined(with: .move(edge: .top)))
            }
        }
        .animation(.easeOut(duration: 0.25), value: store.activity)
    }

    private func listsPanel(_ day: AlmanacDay) -> some View {
        Panel {
            HStack(alignment: .top, spacing: 14) {
                termColumn("", t("common.suits"), day.yi, Palette.gold)
                Divider().overlay(Palette.line)
                termColumn("", t("common.avoid"), day.ji, Palette.rose)
            }
        }
    }

    private func termColumn(_ mark: String, _ label: String, _ terms: [String], _ colour: Color) -> some View {
        VStack(alignment: .leading, spacing: 8) {
            HStack(spacing: 8) {
                Text(l(mark))
                    .font(Typeface.display(20))
                    .foregroundStyle(colour)
                Text(label.uppercased())
                    .font(Typeface.sans(11, weight: .bold))
                    .tracking(1.8)
                    .foregroundStyle(Palette.inkMute)
            }
            ForEach(terms.isEmpty ? ["—"] : terms, id: \.self) { term in
                Text(glossed(term))
                    .font(Typeface.serif(17))
                    .foregroundStyle(Palette.inkSoft)
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
    }

    private func hoursPanel(_ day: AlmanacDay) -> some View {
        Panel(title: t("almanac.hours")) {
            VStack(spacing: 0) {
                ForEach(day.hours) { hour in
                    HStack(spacing: 10) {
                        Text(l(hour.ganzhi))
                            .font(Typeface.display(16))
                            .foregroundStyle(hour.lucky ? Palette.gold : Palette.inkMute)
                            .frame(width: 44, alignment: .leading)
                        Text(hour.range)
                            .font(Typeface.sans(14))
                            .foregroundStyle(Palette.inkSoft)
                            .frame(width: 104, alignment: .leading)
                        Text(l(hour.spirit))
                            .font(Typeface.serif(16))
                            .foregroundStyle(Palette.inkSoft)
                        Spacer()
                        Image(systemName: hour.lucky ? "sun.max.fill" : "moon")
                            .font(.system(size: 12))
                            .foregroundStyle(hour.lucky ? Palette.gold : Palette.inkMute.opacity(0.6))
                    }
                    .padding(.vertical, 7)
                    if hour.id != day.hours.last?.id {
                        Divider().overlay(Palette.line)
                    }
                }
            }
        }
    }

    private func tablesPanel(_ day: AlmanacDay) -> some View {
        Panel(title: t("almanac.tables")) {
            row(l("Stems and branches"), lf("Year {0} · Month {1} · Day {2} · Zodiac {3}", day.lunar.yearGanZhi, day.lunar.monthGanZhi, day.lunar.dayGanZhi, day.lunar.zodiac))
            row(l("Day officer"), l(day.dayOfficer))
            row(l("Lunar mansion"), [day.mansion.name, day.mansion.animal, day.mansion.direction, day.mansion.beast].map(l).joined(separator: " · "))
            row(l("Day spirit"), [day.spirit.name, day.spirit.road, day.spirit.luck].map(l).joined(separator: " · "))
            row(l("Clash and direction"), lf("Clash: {0} · Direction: {1}", day.clash, day.harmDirection))
            if let term = day.solarTerm { row(l("Solar term"), l(term)) }
            else { row(l("Next solar term"), "\(l(day.nextSolarTerm.name)) · \(day.nextSolarTerm.date)") }
            if !day.auspicious.isEmpty { row(l("Favourable spirits"), day.auspicious.map(l).joined(separator: " · ")) }
            if !day.inauspicious.isEmpty { row(l("Unfavourable spirits"), day.inauspicious.map(l).joined(separator: " · ")) }
            if !day.pengzu.isEmpty { row(l("Peng Zu taboos"), day.pengzu.map(l).joined(separator: " · ")) }
        }
    }

    private func row(_ label: String, _ value: String) -> some View {
        HStack(alignment: .top, spacing: 12) {
            Text(l(label))
                .font(Typeface.sans(12, weight: .bold))
                .tracking(1.4)
                .foregroundStyle(Palette.inkMute)
                .frame(width: 76, alignment: .leading)
            Text(l(value))
                .font(Typeface.serif(16))
                .foregroundStyle(Palette.inkSoft)
                .fixedSize(horizontal: false, vertical: true)
            Spacer(minLength: 0)
        }
        .padding(.vertical, 4)
    }

    private func standingWord(_ standing: String) -> String {
        switch standing {
        case "auspicious": t("almanac.yellowDay")
        case "inauspicious": t("almanac.blackDay")
        default: t("almanac.mixedDay")
        }
    }

    private func standingColour(_ standing: String) -> Color {
        switch standing {
        case "auspicious": Palette.gold
        case "inauspicious": Palette.rose
        default: Palette.ink
        }
    }

    private func standingLine(_ day: AlmanacDay) -> String {
        lf("{0} governs the day on the {1}. The day officer is {2}. Clash: {3}; direction: {4}.", day.spirit.name, day.spirit.road, day.dayOfficer, day.clash, day.harmDirection)
    }

    private func judgementLine(_ judgement: AlmanacDay.Judgement, _ day: AlmanacDay) -> String {
        if let matched = judgement.matched {
            return lf(judgement.verdict == "avoid" ? "The almanac lists {0} among activities to avoid." : "The almanac lists {0} among suitable activities.", matched)
        }
        return lf("This activity is not explicitly listed. The day officer is {0}; the day spirit is {1}.", day.dayOfficer, day.spirit.name)
    }

    private func verdictWord(_ verdict: String) -> String {
        switch verdict {
        case "suitable": t("almanac.yes")
        case "avoid": t("almanac.no")
        default: t("almanac.silent")
        }
    }

    private func verdictColour(_ verdict: String) -> Color {
        switch verdict {
        case "suitable": Palette.gold
        case "avoid": Palette.rose
        default: Palette.inkSoft
        }
    }
}

File: native/android/app/src/main/kotlin/art/lazying/auspice/screens/Almanac.kt
package art.lazying.auspice.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ChevronLeft
import androidx.compose.material.icons.filled.ChevronRight
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.*
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.navigation.NavController
import art.lazying.auspice.*
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject

@Composable
fun AlmanacScreen(navController: NavController) {
    var dateText by rememberPracticeState("almanac.date", LocalDate.now().toString())
    val date = LocalDate.parse(dateText)
    var activity by rememberPracticeState<String?>("almanac.activity", null)
    var day by remember { mutableStateOf<AlmanacDay?>(null) }
    var activities by remember { mutableStateOf<List<AlmanacActivity>>(emptyList()) }
    var error by remember { mutableStateOf<String?>(null) }

    LaunchedEffect(date, activity) {
        runCatching {
            if (activities.isEmpty()) {
                activities = Engines.evaluateAs("almanac.activities")
            }
            day = Engines.evaluateAs(
                "almanac.day",
                buildJsonObject {
                    put("date", JsonPrimitive(date.toString()))
                    activity?.let { put("activity", JsonPrimitive(it)) }
                }
            )
            error = null
        }.onFailure { error = l("This reading could not be computed. Please try again.") }
    }

    ScreenScaffold(
        navController,
        eyebrow = t("practice.almanac"),
        title = t("almanac.title"),
        tagline = t("almanac.tagline")
    ) {
        Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
            Icon(
                Icons.Default.ChevronLeft, "Previous day", tint = Palette.inkSoft,
                modifier = Modifier.size(44.dp).clip(CircleShape)
                    .clickable { dateText = date.minusDays(1).toString() }.padding(10.dp)
            )
            Column(Modifier.weight(1f), horizontalAlignment = Alignment.CenterHorizontally) {
                Text(
                    date.format(DateTimeFormatter.ofPattern("EEEE, d MMMM", java.util.Locale.forLanguageTag(Localisation.code))),
                    style = Type.display(18), color = Palette.ink
                )
                day?.let { Text(lunarDateText(it.lunar.text), style = Type.serif(15), color = Palette.inkMute) }
            }
            Icon(
                Icons.Default.ChevronRight, "Next day", tint = Palette.inkSoft,
                modifier = Modifier.size(44.dp).clip(CircleShape)
                    .clickable { dateText = date.plusDays(1).toString() }.padding(10.dp)
            )
        }

        error?.let {
            Panel(title = t("common.notComputed")) { Text(it, style = Type.serif(16), color = Palette.inkSoft) }
        }

        day?.let { d ->
            ExplainReading(Json.encodeToString(d))
            Panel {
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Text(standingWord(d.standing), style = Type.display(26), color = standingColour(d.standing))
                    Spacer(Modifier.width(10.dp))
                    Text(l(d.lunar.dayGanZhi), style = Type.serif(18), color = Palette.inkSoft)
                    Spacer(Modifier.weight(1f))
                    Text(
                        l(d.dayOfficer),
                        style = Type.sans(13, FontWeight.Bold),
                        color = Palette.gold,
                        modifier = Modifier.clip(CircleShape).background(Palette.goldSoft)
                            .padding(horizontal = 10.dp, vertical = 6.dp)
                    )
                }
                Text(
                    lf("{0} governs the day on the {1}. The day officer is {2}. Clash: {3}; direction: {4}.", d.spirit.name, d.spirit.road, d.dayOfficer, d.clash, d.harmDirection),
                    style = Type.serif(17), color = Palette.inkSoft
                )
            }

            Panel(title = t("almanac.canIDoIt")) {
                FlowRowOf {
                    activities.forEach { item ->
                        Chip(l(item.name.en), null, activity == item.id) {
                            activity = if (activity == item.id) null else item.id
                        }
                    }
                }
                d.judgement?.let { judgement ->
                    Column(verticalArrangement = Arrangement.spacedBy(6.dp), modifier = Modifier.padding(top = 4.dp)) {
                        Text(verdictWord(judgement.verdict), style = Type.display(22), color = verdictColour(judgement.verdict))
                        Text(if (judgement.matched != null) lf(if (judgement.verdict == "avoid") "The almanac lists {0} among activities to avoid." else "The almanac lists {0} among suitable activities.", judgement.matched) else lf("This activity is not explicitly listed. The day officer is {0}; the day spirit is {1}.", d.dayOfficer, d.spirit.name), style = Type.serif(17), color = Palette.inkSoft)
                    }
                }
            }

            Panel {
                Row(horizontalArrangement = Arrangement.spacedBy(14.dp)) {
                    TermColumn("", t("common.suits"), d.yi, Palette.gold, Modifier.weight(1f))
                    TermColumn("", t("common.avoid"), d.ji, Palette.rose, Modifier.weight(1f))
                }
            }

            Panel(title = t("almanac.hours")) {
                d.hours.forEachIndexed { index, hour ->
                    Row(Modifier.fillMaxWidth().padding(vertical = 7.dp), verticalAlignment = Alignment.CenterVertically) {
                        Text(l(hour.ganzhi), style = Type.display(16),
                            color = if (hour.lucky) Palette.gold else Palette.inkMute,
                            modifier = Modifier.width(44.dp)
                        )
                        Text(hour.range, style = Type.sans(14), color = Palette.inkSoft, modifier = Modifier.width(112.dp))
                        Text(l(hour.spirit), style = Type.serif(16), color = Palette.inkSoft)
                    }
                    if (index < d.hours.lastIndex) HorizontalDivider(color = Palette.line)
                }
            }

            Panel(title = t("almanac.tables")) {
                TableRow(l("Stems and branches"), lf("Year {0} · Month {1} · Day {2} · Zodiac {3}", d.lunar.yearGanZhi, d.lunar.monthGanZhi, d.lunar.dayGanZhi, d.lunar.zodiac))
                TableRow(l("Day officer"), l(d.dayOfficer))
                TableRow(l("Lunar mansion"), listOf(d.mansion.name, d.mansion.animal, d.mansion.direction, d.mansion.beast).joinToString(" · ") { l(it) })
                TableRow(l("Day spirit"), listOf(d.spirit.name, d.spirit.road, d.spirit.luck).joinToString(" · ") { l(it) })
                TableRow(l("Clash and direction"), lf("Clash: {0} · Direction: {1}", d.clash, d.harmDirection))
                if (d.solarTerm != null) TableRow(l("Solar term"), l(d.solarTerm))
                else TableRow(l("Next solar term"), "${l(d.nextSolarTerm.name)} · ${d.nextSolarTerm.date}")
                if (d.auspicious.isNotEmpty()) TableRow(l("Favourable spirits"), d.auspicious.joinToString(" · ") { l(it) })
                if (d.inauspicious.isNotEmpty()) TableRow(l("Unfavourable spirits"), d.inauspicious.joinToString(" · ") { l(it) })
                if (d.pengzu.isNotEmpty()) TableRow(l("Peng Zu taboos"), d.pengzu.joinToString(" · ") { l(it) })
            }
        }
    }
}

@Composable
private fun TermColumn(mark: String, label: String, terms: List<String>, colour: Color, modifier: Modifier = Modifier) {
    Column(modifier, verticalArrangement = Arrangement.spacedBy(8.dp)) {
        Row(horizontalArrangement = Arrangement.spacedBy(8.dp), verticalAlignment = Alignment.CenterVertically) {
            Text(l(mark), style = Type.display(20), color = colour)
            Text(
                label.uppercase(),
                style = Type.sans(11, FontWeight.Bold).copy(letterSpacing = 1.8.sp),
                color = Palette.inkMute
            )
        }
        (terms.ifEmpty { listOf("—") }).forEach {
            Text(glossed(it), style = Type.serif(17), color = Palette.inkSoft)
        }
    }
}

@Composable
private fun TableRow(label: String, value: String) {
    Row(Modifier.fillMaxWidth().padding(vertical = 4.dp), horizontalArrangement = Arrangement.spacedBy(12.dp)) {
        Text(l(label),
            style = Type.sans(12, FontWeight.Bold).copy(letterSpacing = 1.4.sp),
            color = Palette.inkMute,
            modifier = Modifier.width(80.dp)
        )
        Text(l(value), style = Type.serif(16), color = Palette.inkSoft)
    }
}

private fun standingWord(standing: String) = when (standing) {
    "auspicious" -> t("almanac.yellowDay")
    "inauspicious" -> t("almanac.blackDay")
    else -> t("almanac.mixedDay")
}

private fun standingColour(standing: String) = when (standing) {
    "auspicious" -> Palette.gold
    "inauspicious" -> Palette.rose
    else -> Palette.ink
}

private fun verdictWord(verdict: String) = when (verdict) {
    "suitable" -> t("almanac.yes")
    "avoid" -> t("almanac.no")
    else -> t("almanac.silent")
}

private fun verdictColour(verdict: String) = when (verdict) {
    "suitable" -> Palette.gold
    "avoid" -> Palette.rose
    else -> Palette.inkSoft
}
